### DISCUSSION TITLE
dev-server should rewrite imports for json & css

### DISCUSSION THREAD

**Discussion by jogibear9988** [open-wc/open-wc#2368] [General]
dev-sevrer should rewrite imports for css & json. it should add an assert statement to it
see: https://web.dev/css-module-scripts/

**Comment by daKmoR** [comment]
Why do you think it should do that? what would be the use case?

and why not write the import with the assert into the source directly? 🤔

**Comment by jogibear9988** [comment]
because there are many libs wich have ESM code with css/json imports without asserts, cause they were used via bundlers.
For example monaco-editor of vscode is one of them

**Reply by bennypowers** [reply]
in those cases, end-users should apply bundlers to the dev server, e.g. `rollup-plugin-lit-css`, `rollup-plugin-styles`, etc.

proactively adding import assertions to those imports would break existing apps.

**Reply by jogibear9988** [reply]
why? a import of a css file from javascript in a browser will always result in  an error, if no assert statement is added.

**Reply by justinfagnani** [reply]
The problem is that the bundler semantics are different from the standard semantics. Converting non-standard CSS/JSON imports to standard ones would just break apps.

**Reply by bennypowers** [reply]
moreover, the non-standard `import styles from './styles.css'` might mean different things in different apps. In _my_ app, it might mean "transform the file to a js module with a default export that's a Lit CSSResult", but in someone else's app it might mean "minify and pre-process the CSS, then append a link to the hashed filename to the `<head>`". Because we're talking about non-standard behaviour, there's _no way_ to know in advance which behaviour the user expects.

**Reply by jogibear9988** [reply]
maybe, but when we do nothing, and keep the import as it is, the page will never work. cause javascript in browser could not import it without assert.
how should adding the assert break the app? the import without assert would always fail in the browser. and if the user uses a bundler plugin, the import is transformed, and so the web-dev-server does not need to change it.

**Reply by bennypowers** [reply]
> and if the user uses a bundler plugin, the import is transformed, and so the web-dev-server does not need to change it.

I'm not sure but it seems like there's a miscommunication happening here. WDS's model for this kind of non-standard import is to use bundler plugins to transform files on the fly.

```js
// web-dev-server.config.js
import { fromRollup } from '@web/dev-server-rollup';
import jsonRollup from '@rollup/plugin-json';

const json = fromRollup(jsonRollup);

export default {
  plugins: [json()]
}
```

**Reply by jogibear9988** [reply]
@bennypowers I did understand that. But why can't we add the assert to the css or json import when no bundler plugin is used, or the bundler plugin did not transform. We then know that the import will not work, so we could add.
